### PR TITLE
Dependency locking modes

### PR DESCRIPTION
**Pull request by ljacomet**
Will allow users to define the lock mode under which a build needs to
operate:
* DEFAULT is the way locking worked up to Gradle 6.0
* STRICT is a new mode where the resolution will fail if a dependency is
 marked as locked but does not have a lock state associated.
* LENIENT is a new mode where the lock state will be used to prevent
auto upgrades of version but the validation step is skipped, allowing a
developer to work on the project without having to update the lock state
 after every change impacting resolution

Feature is working and has tests, so already looking for a first feedback.

What's left:
- [x] Prevent changing the lock mode once a first configuration has been resolved
- [x] Documentation

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *     <li>DEFAULT will load the lock state and verify resolution matches it</li>
+ *     <li>STRICT will fail resolution if a locked configuration does not have a lock state</li>
+ *     <li>LENIENT will load the lock state but not perform verification after resolution</li>
+ * </ul>
+ *
+ * @since 6.1
+ */
+@Incubating
+public
+enum LockMode {
```
Comment (melix) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java]: :o: Weird layout :)

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java [hunk 2]
```diff
@@ -103,6 +105,25 @@
...
     ConfigurationContainer getConfigurations();
 
+    /**
+     * Configures dependency locking
+     *
+     * @param configureClosure the configuration action
+     * @since 6.1
+     */
+    @Incubating
+    void dependencyLocking(Closure configureClosure);
```
Comment (melix) [inline subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java]: :o: Do you need a `Closure` here? I think for new things you can just use `Action` and let Gradle do the decoration, no?

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package org.gradle.api.artifacts.dsl;
+
+import org.gradle.api.Incubating;
+
+/**
+ * The supported lock modes:
+ * <ul>
+ *     <li>DEFAULT will load the lock state and verify resolution matches it</li>
+ *     <li>STRICT will fail resolution if a locked configuration does not have a lock state</li>
+ *     <li>LENIENT will load the lock state but not perform verification after resolution</li>
```
Comment (melix) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java]: :x: The first reaction I have reading this, is, then why do you load the lock state? It feels like this section deserves a bit more explanation on the use cases. In particular, explain _what_ validation is done and not done.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java]: Indeed :-)

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java (hunk 2 above)
Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java]: Looks like I do need a `Closure`. I first tried with an `Action` but then:
```
buildscript {
    dependencyLocking {
        lockMode = LockMode.LENIENT
    }
}
```
was invalid in a Groovy build script.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java (hunk 3 above)
Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/LockMode.java]: Right, the description of this PR has a more complete explanation. Will improve the javadoc here as well.

**Comment by ljacomet** [comment]
@melix The reason I am not adding a command line flag is that I do not think that's the recommended usage pattern.
What I envision instead is team based decisions on when you go `LENIENT`: in composites, when a property is set in `gradle.properties`, etc ...
It is not something you use for a _single_ build IMO.
Makes sense?

**Comment by ljacomet** [comment]
@melix I have investigated implementing the lock mode per configuration resolution strategy. It complicates the code somewhat, because of the relationship between a configuration and the creation of the root local metadata. So I gave up for now.
This remains an option, but since I do not see a valid use case for having different modes in parallel, I am not doing it now.

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java (hunk 2 above)
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java]: I guess it means that hte script handler is not created via an instantiator.

**Pull request merged**